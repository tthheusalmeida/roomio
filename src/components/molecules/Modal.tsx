"use client";
import { mergeClassNames } from "@/utils/classNames";
import { useEffect } from "react";
import { IoMdClose } from "react-icons/io";

interface ModalProps {
  isOpen: boolean;
  expanded?: boolean;
  title?: string;
  children: React.ReactNode;
  showCloseButton?: boolean;
  onClose?: () => void;
}

export default function Modal({
  isOpen,
  expanded,
  title,
  children,
  showCloseButton = false,
  onClose,
}: ModalProps) {
  useEffect(() => {
    if (isOpen) {
      document.body.classList.add("overflow-hidden");
    } else {
      document.body.classList.remove("overflow-hidden");
    }

    return () => {
      document.body.classList.remove("overflow-hidden");
    };
  }, [isOpen]);

  return (
    <div
      className={`fixed inset-0 z-50 flex items-end justify-center transition-opacity duration-200 bg-[var(--color-background)]/50 ${
        isOpen
          ? "opacity-100 pointer-events-auto"
          : "opacity-0 pointer-events-none"
      }`}
    >
      <div
        className={mergeClassNames(`
          relative w-full max-h-full bg-[var(--color-foreground)] shadow-xl overflow-y-auto
          transition-all duration-600 ease-in-out
          transform safe-padding-bottom
          ${isOpen ? "translate-y-0 opacity-100" : "translate-y-full opacity-0"}
          ${expanded ? "h-full" : "rounded-t-xl"}
        `)}
      >
        {/* Title */}
        {title && (
          <div className="mx-4 my-5 flex justify-between items-center">
            <h2 className="text-sm font-semibold text-gray-700">{title}</h2>

            {showCloseButton && (
              <IoMdClose
                size={24}
                onClick={onClose}
                className="cursor-pointer"
              />
            )}
          </div>
        )}

        {/* Content */}
        <div className="flex flex-col flex-1 m-4">{children}</div>
      </div>
    </div>
  );
}
