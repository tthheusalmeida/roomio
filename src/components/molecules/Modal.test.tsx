import { render, screen, fireEvent } from "@testing-library/react";
import Modal from "./Modal";

describe("Modal component", () => {
  beforeEach(() => {
    document.body.className = "";
  });

  it("renders children correctly", () => {
    render(
      <Modal isOpen={true}>
        <div data-testid="modal-content">Modal Content</div>
      </Modal>
    );

    const content = screen.getByTestId("modal-content");
    expect(content).toBeInTheDocument();
  });

  it("adds overflow-hidden to body when open", () => {
    const { rerender } = render(<Modal isOpen={true}>Content</Modal>);
    expect(document.body).toHaveClass("overflow-hidden");

    rerender(<Modal isOpen={false}>Content</Modal>);
    expect(document.body).not.toHaveClass("overflow-hidden");
  });

  it("renders title when provided", () => {
    render(
      <Modal isOpen={true} title="My Modal">
        Content
      </Modal>
    );
    expect(screen.getByText("My Modal")).toBeInTheDocument();
  });

  it("renders close button and calls onClose when clicked", () => {
    const handleClose = jest.fn();
    render(
      <Modal isOpen={true} title="Modal" showCloseButton onClose={handleClose}>
        Content
      </Modal>
    );

    const icon = document.querySelector(".cursor-pointer") as HTMLElement;
    expect(icon).toBeInTheDocument();

    fireEvent.click(icon);
    expect(handleClose).toHaveBeenCalled();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Modal isOpen={true} title="Snapshot Modal" showCloseButton>
        <div>Content</div>
      </Modal>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
