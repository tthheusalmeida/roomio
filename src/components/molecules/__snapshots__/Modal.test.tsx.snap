// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Modal component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 z-50 flex items-end justify-center transition-opacity duration-200 bg-[var(--color-background)]/50 opacity-100 pointer-events-auto"
  >
    <div
      class="relative w-full max-h-full bg-[var(--color-foreground)] shadow-xl overflow-y-auto transition-all duration-600 ease-in-out transform safe-padding-bottom translate-y-0 opacity-100 rounded-t-xl"
    >
      <div
        class="mx-4 my-5 flex justify-between items-center"
      >
        <h2
          class="text-sm font-semibold text-gray-700"
        >
          Snapshot Modal
        </h2>
        <svg
          class="cursor-pointer"
          fill="currentColor"
          height="24"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M405 136.798L375.202 107 256 226.202 136.798 107 107 136.798 226.202 256 107 375.202 136.798 405 256 285.798 375.202 405 405 375.202 285.798 256z"
          />
        </svg>
      </div>
      <div
        class="flex flex-col flex-1 m-4"
      >
        <div>
          Content
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
